### DISCUSSION TITLE
alternativa para logs/trace espalhados por todo moclojer-app

### DISCUSSION THREAD

**Discussion by Felipe-gsilva** [moclojer/moclojer-app#369] [Ideas]
foi discudido no servidor do discord criar uma forma de "reescrever" as partes do código que utilizam o logs/trace como descrito nessa branch: https://github.com/moclojer/moclojer-app/blob/refactor/telemere/src/back/api/controllers/mocks.clj#L169

**Comment by Felipe-gsilva** [comment]
eu e o @J0sueTM conversamos em call e percebemos uma insconstância na ideia atual de envolver as funções de x namespace por sua versão reescria com o trace. No caso abaixo, funções que recebem um argumento 'descontruido' (vide {:keys [a b c]}, a função atual não retornaria os values do mapa, apenas a string "{:a a :b :c c}".


```clj
(doseq [[fsym func] (ns-publics *ns*)]
  (let [args (first (:arglists (meta func)))
        argmap (reduce
                (fn [acc v]
                  (assoc acc (keyword v) v))
                ;; remove components and ctx
                {} (drop-last 2 args))
        symkey (keyword (symbol (str *ns*) (str fsym)))]
    (intern
     *ns* fsym
     (eval
      `(fn [~@args]
         (com.moclojer.components.logs/trace
          ~symkey ~argmap
          (~func ~@args)))))))

```

@avelino @matheusfrancisco pensamos em reescrever todas as funções sem a desconstrução, ou seja, só passar os args e qualquer extra seria passando usando o operador & (o que é um trabalho muito grande e não otimizado)

a outra ideia seria criar um parser do mapa retornado em string, o que parece ser bem mais rápido de ser feito. O que vocês acham?

**Reply by J0sueTM** [reply]
@Felipe-gsilva acredito que parsear as desconstrucoes seja melhor. Vou fazer um prototipo e mando aqui jaja

**Reply by Felipe-gsilva** [reply]
@J0sueTM arrumei os erros da função que você tinha mandado, mas ainda faltam os casos onde a função recebe a keyword :or, pra valores "default"

vou tentar entender como fazer isso, mas fiz um protótipo n testado dela

**Comment by Felipe-gsilva** [comment]
@J0sueTM @avelino @matheusfrancisco 
pessoal, terminei de fazer um template daquela funcao de colocar todas as funcoes que queremos com trace e queria saber se vcs gostam do jeito q ela esta no momento

obs: ainda estou tendo um erro com os args '&', mas acredito que possamos muda-los para o {:keys } como haviamos comentado anteriormente

```clj
(defn trace-all-ns
  "Iterate over all desired ns functions and replace them with themselves within
  a `logs/trace` call that uses each function's arglist as context."
  []
  (let [desired-ns (->> (all-ns)
                        (map ns-name)
                        (map name)
                        (filter #(clojure.string/starts-with? % "com.moclojer"))
                        (filter #(not (clojure.string/ends-with? % "logs")))
                        (map symbol)
                        (into []))]
    (doseq [current-ns desired-ns]
      (clojure.pprint/pprint current-ns)
      (doseq [[fsym func] (ns-publics current-ns)]
        (let [symkey (keyword (symbol (str current-ns) (name fsym)))
              args (first (:arglists (meta func)))
              pre-argmap (fn [acc v]
                           (cond
                     ;; Arg with a :as alias, making it not necessary to
                     ;; get the inner deconstructed keys.
                             (and (map? v) (:as v))
                             (acc (keyword (:as v)) (:as v))
                     ;; Arg with only the :keys vec declared.
                             (and (map? v) (:keys v))
                             (into {} (map (fn [k]
                                             [(keyword k) k])
                                           (:keys v)))
                     ;; Normal arg types.
                             :else (assoc acc (keyword v) v)))
              argmap (reduce pre-argmap {} (drop-last 2 args))
              callargs (map
                        (fn [arg]
                          (cond
                            (and (map? arg) (:as arg))
                            (:as arg)
                            (and (map? arg) (:keys arg))
                            (into {} (map (fn [a]
                                            [(keyword a) a])
                                          (:keys arg)))
                            :else arg))
                        args)]
          (intern
                    *ns* fsym
                    (eval `(fn [~@args]
                             (com.moclojer.components.logs/trace ~symkey ~argmap
                                                                 (~func ~@callargs))))))))))
```

**Reply by Felipe-gsilva** [reply]
Essas são as funções que estão sendo colocadas com trace e abaixo as com o erro do &:

; (out) #'back.api.controllers.mocks/map->Database
; (out) #'back.api.controllers.mocks/->Database
; (out) #'back.api.controllers.mocks/execute
; (out) #'back.api.controllers.mocks/build-db-logger
; (out) #'back.api.controllers.mocks/DatabaseProvider
; (out) #'back.api.controllers.mocks/version
; (out) #'back.api.controllers.mocks/close-client
; (out) #'back.api.controllers.mocks/create-client
; (out) #'back.api.controllers.mocks/->MQMock
; (out) #'back.api.controllers.mocks/op-name->op-fn
; (out) #'back.api.controllers.mocks/mock-op-name->op-fn
; (out) #'back.api.controllers.mocks/wrap-worker
; (out) #'back.api.controllers.mocks/mock-channels
; (out) #'back.api.controllers.mocks/map->MQMock
; (out) #'back.api.controllers.mocks/try-op!
; (out) #'back.api.controllers.mocks/IMQ
; (out) #'back.api.controllers.mocks/start-job!
; (out) #'back.api.controllers.mocks/map->MQ
; (out) #'back.api.controllers.mocks/->MQ
; (out) #'back.api.controllers.mocks/to-jdbc-uri
; (out) #'back.api.controllers.mocks/pack-workers-channels
; (out) #'back.api.controllers.mocks/unarquive-channel!
; (out) #'back.api.controllers.mocks/create-listener
; (out) #'back.api.controllers.mocks/group-handlers-by-channel
; (out) #'back.api.controllers.mocks/->Sentry
; (out) #'back.api.controllers.mocks/SentryLogger
; (out) #'back.api.controllers.mocks/set-as-default-exception-handler
; (out) #'back.api.controllers.mocks/map->SentryMock
; (out) #'back.api.controllers.mocks/map->Sentry
; (out) #'back.api.controllers.mocks/send-event!
; (out) #'back.api.controllers.mocks/->SentryMock
; (out) #'back.api.controllers.mocks/pack-pattern
; (out) #'back.api.controllers.mocks/unpack-pattern

; (out) could not convert #'com.moclojer.rq.queue/pop! and [client queue-name & options]
; (out) could not convert #'com.moclojer.rq.queue/llen and [client queue-name & options]
; (out) could not convert #'com.moclojer.rq.queue/push! and [client queue-name message & options]
; (out) could not convert #'com.moclojer.rq.pubsub/publish! and [client channel message & options]
; (out) could not convert #'com.moclojer.rq.pubsub/subscribe! and [client workers & options]
